src/sentry/auth/staff.py

```diff
@@ -0,0 +1,319 @@
+        return True
+    stf = getattr(request, "staff", None) or Staff(request)
+    return stf.is_active
+
+
+class Staff(ElevatedMode):
+    allowed_ips = frozenset(ipaddress.ip_network(str(v), strict=False) for v in ALLOWED_IPS)
+
+    def __init__(self, request, allowed_ips=UNSET, current_datetime=None):
+        self.uid: str | None = None
+        self.request = request
+        if allowed_ips is not UNSET:
+            self.allowed_ips = frozenset(
+                ipaddress.ip_network(str(v), strict=False) for v in allowed_ips or ()
+            )
+        self._populate(current_datetime=current_datetime)
+
+    @property
+    def is_active(self):
+        # We have a wsgi request with no user.
+        if not hasattr(self.request, "user"):
+            return False
+        # if we've been logged out
+        if not self.request.user.is_authenticated:
+            return False
+        # if staff status was changed
+        if not self.request.user.is_staff:
+            return False
+        # if the user has changed
+        if str(self.request.user.id) != self.uid:
+            return False
+        return self._is_active
+
+    def is_privileged_request(self):
+        """
+        Returns ``(bool is_privileged, str reason)``
+        """
+        allowed_ips = self.allowed_ips
+
+        # _admin should have always completed SSO to gain status.
+        # We expect ORG_ID to always be set in production.
+        if ORG_ID and not has_completed_sso(self.request, ORG_ID):
+            if not DISABLE_SSO_CHECK_FOR_LOCAL_DEV:
+                return False, "incomplete-sso"
+
+        # if there's no IPs configured, we allow assume its the same as *
+        if not allowed_ips:
+            return True, None
+        ip = ipaddress.ip_address(str(self.request.META["REMOTE_ADDR"]))
+        if not any(ip in addr for addr in allowed_ips):
+            return False, "invalid-ip"
+        return True, None
+
+    def get_session_data(self, current_datetime=None):
+        """
+        Return the current session data, with native types coerced.
+        """
+        request = self.request
+        data = request.session.get(SESSION_KEY)
+
+        try:
+            cookie_token = request.get_signed_cookie(
+                key=COOKIE_NAME,
+                default=None,
+                salt=COOKIE_SALT,
+                max_age=MAX_STAFF_SESSION_AGE.total_seconds(),
+            )
+        except BadSignature:
+            logger.exception(
+                "staff.bad-cookie-signature",
+                extra={"ip_address": request.META["REMOTE_ADDR"], "user_id": request.user.id},
+            )
+            return
+
+        if not cookie_token:
+            if data:
+                logger.warning(
+                    "staff.missing-cookie-token",
+                    extra={"ip_address": request.META["REMOTE_ADDR"], "user_id": request.user.id},
+                )
+            return
+        elif not data:
```

Comment:
Is this conditional supposed to be mutually exclusive? Currently we check for `data` being empty only if `cookie_token` is present. Should we be checking for data presence in general?